File: populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None: #Tree does not exist
            return None
        
        current_level = 1
        next_level = 0
        queue = [root]
        while queue: # BFS
            node = queue.pop(0)
            if node.left: # Add left child if it exists
                queue.append(node.left)
                next_level += 1
            if node.right: # Add right child if it exists
                queue.append(node.right)
                next_level += 1
            if current_level != 1: # Not on edge of tree level
                node.next = queue[0]

            current_level -= 1
            if current_level == 0: #Moved to next level of tree
                current_level = next_level
                next_level = 0
        
        return root

File: populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution 
{
    public Node connect(Node root) 
    {
        if (root == null) // No tree
            return root;
        
        // Create queue for BFS
        Deque<Node> queue = new LinkedList<Node>();
        queue.addLast(root);
        int currentLevel = 1, nextLevel = 0;
        while (queue.size() > 0) // BFS
        {
            Node node = queue.removeFirst();
            currentLevel--;
            
            if (node.left != null) // Check left
            {
                queue.addLast(node.left);
                nextLevel++;
            }
            if (node.right != null) // Check right
            {
                queue.addLast(node.right);
                nextLevel++;
            }
            
            if (currentLevel == 0) // Check next
            {
                currentLevel = nextLevel;
                nextLevel = 0;
            }
            else // Add next
                node.next = queue.peekFirst();
        }
        return root;
    }
}